### PR TITLE
Ultra-Light Check-In

### PR DESCRIPTION
**Pull request by kruai**
This PR is meant to address [RHCLOUD-9412](https://issues.redhat.com/browse/RHCLOUD-9412).

This adds a PUT endpoint that accepts a list of canonical facts and a time offset (in minutes). Given the list of canonical facts, it should find one specific host, and update its staleness timestamp using the supplied time offset.

### CODE REVIEW

**Comment by tahmidefaz** [comment]
@kruai since we are only updating a portion of the host data, I am thinking a `PATCH` maybe the correct method to use rather than a `POST`?

**Comment by thearifismail** [comment]
@kruai tests?

**Comment by kruai** [comment]
> @kruai since we are only updating a portion of the host data, I am thinking a `PATCH` maybe the correct method to use rather than a `POST`?

There was a lot of discussion on this, actually; the general consensus was to make it a `POST`. I think the reasoning was that we're adding a new check-in/heartbeat. 

> @kruai tests?
I'll add those once I actually have finished code lmao

**Review by jharting — CHANGES_REQUESTED** [review]
@kruai don't forget to emit an `updated` event for any host state change

**Comment by kruai** [comment]
> @kruai don't forget to emit an `updated` event for any host state change

Didn't realize that was a thing! Will do.



> is there a test attempting to cull a host with timedelta < 14 days?

There are so many tests it's hard for me to answer this, but I also don't know if that's relevant to this PR! Unless you're asking if I've added tests for my check-in endpoint that set the culling timestamp at less than 14 days... to which the answer is no

**Comment by lphiri** [comment]
@jharting @kruai  - how would this behave if the canonical facts passed in did not match any hosts?
Would this create a new host entry?  I am assuming we do not want this to be backdoor for adding hosts via the REST API,
which was recently removed.

**Comment by kruai** [comment]
> @jharting @kruai - how would this behave if the canonical facts passed in did not match any hosts?
> Would this create a new host entry? I am assuming we do not want this to be backdoor for adding hosts via the REST API,
> which was recently removed.

At the moment, this scenario errors out with my code. I can add a check to return a 400 response if a matching host isn't found. Does that work?

**Comment by lphiri** [comment]
> > @jharting @kruai - how would this behave if the canonical facts passed in did not match any hosts?
> > Would this create a new host entry? I am assuming we do not want this to be backdoor for adding hosts via the REST API,
> > which was recently removed.
> 
> At the moment, this scenario errors out with my code. I can add a check to return a 400 response if a matching host isn't found. Does that work?

@kruai yes, returning a 400 works

**Comment by kruai** [comment]
I need to debug to figure out why my my new test's requests are returning 400. I think it might have something to do with the Insights ID or Account atm

**Review by tahmidefaz — CHANGES_REQUESTED** [review]
In addition to the test, I also have the following questions/suggestions:

* When I sent out the following invalid JSON object, I get a `500`. I think it should throw a `4xx` or something else
```
{
	"canonical_facts": {"insights_id": "3f01b55457674041b75e41829bcee1dc"}
}
```
* When I sent a request with `staleness_offset: 60`, the `staleness_warning_timestamp` shrunk, but the `culled_timestamp` grew to 1 hour in the future. Is this expected behavior?

File: tests/test_api_hosts_update.py [hunk 1]
```diff
@@ -42,6 +46,29 @@ def test_update_fields(patch_doc, event_producer_mock, db_create_host, db_get_ho
...
+    expected_stale_timestamp = host.stale_timestamp + timedelta(minutes=staleness_offset)
+    response_status, response_data = api_put(build_host_checkin_url(), put_doc)
+
+    assert_response_status(response_status, expected_status=201)
+    record = db_get_host(created_host.id)
+
+    expected_stale_timestamp = original_timestamp + timedelta(minutes=staleness_offset)
+
+    assert record.stale_timestamp == expected_stale_timestamp
+
```
[the remark this answers is not in the document]
Comment (kruai) [reply tests/test_api_hosts_update.py:70]: For sure! I'll re-request your review once I add that test and make some fixes. Thank you

**Comment by kruai** [comment]
> * When I sent a request with `staleness_offset: 60`, the `staleness_warning_timestamp` shrunk, but the `culled_timestamp` grew to 1 hour in the future. Is this expected behavior?

I'm actually not sure what people are expecting for the other timestamps tbh. My goal was to push only the `stalness_timestamp` back, but I'm not aware of how this should affect other timestamps. How have these been handled in the past @tahmidefaz ?

**Comment by ryandillinfelton** [comment]
> > * When I sent a request with `staleness_offset: 60`, the `staleness_warning_timestamp` shrunk, but the `culled_timestamp` grew to 1 hour in the future. Is this expected behavior?
> 
> I'm actually not sure what people are expecting for the other timestamps tbh. My goal was to push only the `stalness_timestamp` back, but I'm not aware of how this should affect other timestamps. How have these been handled in the past @tahmidefaz ?

The other timestamps are just a fixed offset away from the base `staleness_timestamp`. You can create the whole set of updated timestamps using the base one and the functions defined [here](https://github.com/RedHatInsights/insights-host-inventory/blob/master/app/culling.py#L25). Here's an [example of them being updated](https://github.com/RedHatInsights/insights-host-inventory/blob/master/app/queue/queue.py#L101).

File: api/host.py [hunk 2]
```diff
@@ -438,3 +442,32 @@ def _build_serialized_tags(host_list, search):
...
     json_output = build_collection_response(tags_list, page, per_page, total)
     return flask_json_response(json_output)
+
+
+@api_operation
+@rbac(Permission.WRITE)
+@metrics.api_request_time.time()
+def host_checkin(body):
+    facts = body.get("canonical_facts")
+    staleness_offset = body.get("staleness_offset")
```
Comment (gmcculloug) [inline api/host.py:452]: @kruai I'm thinking this parameter should be something like `checkin_frequency` for the caller.  If we use it to determine the `staleness_offset` would be an internal decision for the app.  At the moment we may not use it to adjust to staleness_timestamp, but instead to keep track of the next expected checkin.  This allows a caller to use that information to determine how late a system is for a checking in.

Thoughts?

Comment (kruai) [reply api/host.py:452]: That works for me, especially if this value is also going to affect the other timestamps.

**Comment by kruai** [comment]
> > > * When I sent a request with `staleness_offset: 60`, the `staleness_warning_timestamp` shrunk, but the `culled_timestamp` grew to 1 hour in the future. Is this expected behavior?
> > 
> > 
> > I'm actually not sure what people are expecting for the other timestamps tbh. My goal was to push only the `stalness_timestamp` back, but I'm not aware of how this should affect other timestamps. How have these been handled in the past @tahmidefaz ?
> 
> The other timestamps are just a fixed offset away from the base `staleness_timestamp`. You can create the whole set of updated timestamps using the base one and the functions defined [here](https://github.com/RedHatInsights/insights-host-inventory/blob/master/app/culling.py#L25). Here's an [example of them being updated](https://github.com/RedHatInsights/insights-host-inventory/blob/master/app/queue/queue.py#L101).

It took me a while to realize that only the staleness_timestamp is stored, and the other 2 timestamps are only generated while outputting (during serialization). I've changed it up so that I modify the staleness_timestamp and use the default config for output

File: tests/test_api_hosts_update.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (kruai) [reply tests/test_api_hosts_update.py:68]: I assumed that defining a default value in `api.spec.yml` would make it supply the value in `host.py`, but I guess that was wrong. I'm now defaulting the value of checkin_frequency if it isn't passed in

**Review by tahmidefaz — CHANGES_REQUESTED** [review]
I think there's still an issue with the timestamp fields. 
I had a host with the following timestamps:
```
"stale_timestamp": "2020-11-03T06:20:08+00:00",
"stale_warning_timestamp": "2020-11-10T06:20:08+00:00",
"culled_timestamp": "2020-11-17T06:20:08+00:00",
```
Then, I updated it with `staleness_offset: 60`. This updated the host to have: 
```
"stale_timestamp": "2020-11-03T07:20:08+00:00",
"stale_warning_timestamp": "2020-11-03T08:20:08+00:00",
"culled_timestamp": "2020-11-17T07:20:08+00:00",
```
The `culled_timestamp` value looks good, however, the `stale_warning_timestamp` is incorrect. I think it should be `"2020-11-10T07:20:08+00:00"`. Can you repro this? I am not sure where exactly you need to make this change. You might need to do some digging :slightly_smiling_face:

**Review by tahmidefaz — APPROVED** [review]
Ok, I think these changes are LGTM. However, I would feel much comfortable with at least 1 more :heavy_check_mark: before merging :smile:

**Comment by kruai** [comment]
@jharting @ryandillinfelton @subpop @gravitypriest @lphiri @iphands: Any of you wanna weigh in before it's merged?

**Comment by gmcculloug** [comment]
@jharting @subpop @gravitypriest Would really appreciate getting your review on this PR.  Thanks

**Comment by subpop** [comment]
I don't feel I can give an overall LGTM as I am not familiar with the project codebase, but the API interface that the client will be interacting with LGTM.

**Comment by thearifismail** [comment]
> There are so many tests it's hard for me to answer this, but I also don't know if that's relevant to this PR! Unless you're asking if I've added tests for my check-in endpoint that set the culling timestamp at less than 14 days... to which the answer is no

IMO, this test is relevant.  Since its relevance spreads beyond this PR, this PR should be not be blocked.

**Review by thearifismail — APPROVED** [review]
LGTM.  All tests executed successfully.

**Comment by kruai** [comment]
I don't have the option to dismiss @jharting 's review requesting changes (or the ability to re-request a review), so I'm going to need @tahmidefaz or @ryandillinfelton to merge this

**Review by iphands — APPROVED** [review]
Testing

**Review by ryandillinfelton — APPROVED** [review]
LGTM

**Comment by kruai** [comment]
@ryandillinfelton @tahmidefaz I'm not allowed to merge it... can you?
Or do we have to wait for @jharting to update his review?

**Comment by gmcculloug** [comment]
@iphands The other week you were thinking that the `staleness` timestamp should just be bumped by the default internal (1 day) during the ultra-light check-in instead of the check-in frequency value.  Is that still what you are thinking you want?

If I'm reading the code, and PR description, correctly we are still using the [check-in frequency with an default of 1 day]( https://github.com/RedHatInsights/insights-host-inventory/pull/720/files#diff-a5c55b46cc9c536070cb56798c97adfaR452).

I had made reference to this previously in https://github.com/RedHatInsights/insights-host-inventory/pull/720#discussion_r500647020 which is easily missed.

File: tests/test_api_hosts_update.py [hunk 3]
```diff
@@ -42,6 +46,41 @@ def test_update_fields(patch_doc, event_producer_mock, db_create_host, db_get_ho
...
+    put_doc = {
+        "canonical_facts": {"insights_id": f"{created_host.canonical_facts['insights_id']}"},
+        "checkin_frequency": checkin_frequency,
+    }
+
+    response_status, response_data = api_put(build_host_checkin_url(), put_doc)
+
+    assert_response_status(response_status, expected_status=201)
+    record = db_get_host(created_host.id)
+    expected_stale_timestamp = original_timestamp + timedelta(minutes=checkin_frequency)
```
Comment (jharting) [inline tests/test_api_hosts_update.py:65]: I don't think this is correct. The stale_timestamp should not be set to previous stale_timestamp + checkin_frequency. e.g. if stale_timestamp is set to 1 month into the future and checkin_frequence is 60 minutes the outcome shouldn't be 1 month + 60 minutes. Instead, stale_timestamp should be set to now() + 60 minutes, i.e. if the host does not do another "ultralight" check-in 60 minutes from now it is considered stale.

File: api/host.py [hunk 4]
```diff
@@ -440,3 +442,27 @@ def _build_serialized_tags(host_list, search):
...
+@rbac(Permission.WRITE)
+@metrics.api_request_time.time()
+def host_checkin(body):
+    facts = body.get("canonical_facts")
+    staleness_offset = timedelta(minutes=body.get("checkin_frequency") or 1440)
+    serialized_host, host_id, insights_id, updated = update_host_staleness(
+        current_identity.account_number, facts, staleness_offset
+    )
+    if serialized_host:
+        _emit_patch_event(serialized_host, host_id, insights_id)
```
Comment (jharting) [inline api/host.py:457]: The produced event is not correct. It is missing host tags, system_profile and request_id information.

**Review by jharting — CHANGES_REQUESTED** [review]
I don't think that the implemented logic matches the expected behavior (see inline comment).

Moreover, I did not find any test coverage for the updated event emitted on host checkin. Manual testing revealed problems with the event payload.

File: lib/host_repository.py [hunk 5]
```diff
@@ -57,6 +59,25 @@ def add_host(input_host, staleness_offset, update_system_profile=True, fields=DE
...
 
+def update_host_staleness(account_number, canonical_facts, staleness_offset):
+    """
+    Updates the staleness timestamp for a host with matching canonical facts.
+    """
+
+    existing_host = find_existing_host(account_number, canonical_facts)
+    if existing_host:
+        input_host = Host(
+            {"insights_id": existing_host.canonical_facts["insights_id"]},
```
Comment (jharting) [inline lib/host_repository.py:70]: what if the host does not have insights_id?

Comment (kruai) [reply lib/host_repository.py:70]: Oh, I thought they all had to have `insights_id`. Guess I'll need to revise this

File: api/host.py (hunk 4 above)
Comment (kruai) [reply api/host.py:457]: I'm not sure what I'm doing wrong here... I'm using the function in the exact same way as `patch_by_id`. Looking into it, though

File: tests/test_api_hosts_update.py (hunk 3 above)
Comment (kruai) [reply tests/test_api_hosts_update.py:65]: Oh, okay. I can change it to do that

Comment (kruai) [reply tests/test_api_hosts_update.py:65]: Updated

File: lib/host_repository.py (hunk 5 above)
Comment (kruai) [reply lib/host_repository.py]: I'm now passing in the existing canonical facts instead of assuming that `insights_id` exists.

File: api/host.py (hunk 4 above)
Comment (kruai) [reply api/host.py:457]: The only thing I could think of is to set "update_system_profile" to true, but I'm not sure how to test the produced events yet. Got any idea what I could be missing here?

File: lib/host_repository.py [hunk 6]
```diff
@@ -57,6 +61,25 @@ def add_host(input_host, staleness_offset, update_system_profile=True, fields=DE
...
 
 
+def update_host_staleness(account_number, canonical_facts, staleness_offset):
+    """
+    Updates the staleness timestamp for a host with matching canonical facts.
+    """
+
+    existing_host = find_existing_host(account_number, canonical_facts)
+    if existing_host:
+        input_host = Host(
```
Comment (jharting) [inline lib/host_repository.py:71]: I think that rather than creating new Host() and then calling `update_existing_host` (which is used to update the entire host on upload) it would be better to do

```
existing_host._update_stale_timestamp()
db.session.commit()
```
i.e. update the field(s) you need and commit the session.

Then, not unlike in the PATCH case, serialize the representation and emit the event
https://github.com/RedHatInsights/insights-host-inventory/blob/master/api/host.py#L306

Comment (jharting) [reply lib/host_repository.py:71]: Note that it might be a good idea to define a specific reporter for these ultralight checkins. (As otherwise we risk confusion as of why this host with reporter yupana has stale_timestamp that yupana would never set?)

Comment (kruai) [reply lib/host_repository.py:71]: OK... What value makes sense for this new reporter value, then? Should I just make one up, and call it "heartbeat" or "checkin" or something? I'd like @gmcculloug and @iphands opinions on this one

Comment (jharting) [reply lib/host_repository.py:71]: Yes, I think we should define a new reporter value. Either of the examples you used looks good to me.

Comment (kruai) [reply lib/host_repository.py]: I've made the requested changes. Afterwards, the function in host_repository didn't seem necessary, so I've consolidated the code into the function in host.py.
I'm marking this as "resolved" to show that I've addressed these concerns, but please un-resolve it if there's something I missed.

File: api/host.py (hunk 4 above)
Comment (kruai) [reply api/host.py]: It now functions nearly identically to the PATCH call. Can you please un-resolve this if you're still seeing the issue?

**Review by jharting — APPROVED** [review]
Functionality-wise looks good to me. I made a few improvements to the tests.

**Comment by kruai** [comment]
Awesome, thank you!! I'll resolve the merge conflict(s) and merge.

File: swagger/api.spec.yaml [hunk 7]
```diff
@@ -714,6 +751,24 @@ components:
...
+      title: Check-in data
+      description: >-
+        Data required to create a check-in record for a host.
+      type: object
+      required:
+        - canonical_facts
+      properties:
+        canonical_facts:
+          description: A set of canonical_facts belonging to the host.
+          $ref: '#/components/schemas/Facts'
```
Comment (Glutexo) [inline swagger/api.spec.yaml:764]: This is not _canonical facts_, but _facts_.

**Review by Glutexo — COMMENTED** [review]
I found a bug in the schema. There is a field that should contain canonical facts, but it’s schema is defined as Host facts. The fact that this went unnoticed also points to a fact that this is not tested.

**Review by Glutexo — COMMENTED** [review]
One more comment to the schema. The new _CreateCheckIn_ schema is the only place in the 
whole specification where Canonical Facts is explicitly mentioned. We don’t expose this concept; for the outside world, these are first class properties of a Host object. Thus I don’t think it is correct to break this concept in the new endpoint.

I’d say, just put them in the root CreateCheckIn schema. _Flat is better than nested._ I personally like nested more, and even here, we could just call it like _host_ instead of _canonical_facts_, but it would be inconsistent.

```yaml
CreateCheckIn:
  …
  properties:
    insights_id: …
    suscription_manager_id: …
    …
    checkin_frequency: …
```

Of course, my suggestion to DRY the Canonical Facts property set to an extracted schema and use it here still applies. It would require to use the _allOf_ JSON Schema combinator.

**Review by Glutexo — COMMENTED** [review]
Is it correct to return _400 Bad Request_ when there is no matching host found? I’d expect _404 Not Found_ instead. I thought that _400 Bad Request_ should occur only if the request is malformed.

Gave this a second thought after reading https://stackoverflow.com/a/44915434/1307676. That use case is however slightly different from ours. Trying to wrap my head around our check-in, I imagine that the request consists of two parts: ⑴ matching the host and ⑵ writing the check-in data. No matter whether we consider this to be a new Check-in resource operation, or a direct Host update, it could be split into two requests: ⑴ would match the host and either redirect or bail out with _404 Not Found_. ⑵ would return _200 OK_ or _201 Created_ depending on the semantics.

File: api/host.py [hunk 8]
```diff
@@ -441,3 +445,29 @@ def _build_serialized_tags(host_list, search):
...
+    facts = body.get("canonical_facts")
+    staleness_offset = timedelta(minutes=body.get("checkin_frequency") or 1440)
+    existing_host = find_existing_host(current_identity.account_number, facts)
+
+    if existing_host:
+        existing_host._update_stale_timestamp(datetime.now(timezone.utc) + staleness_offset, "checkin")
+        db.session.commit()
+        serialized_host = serialize_host(existing_host, staleness_timestamps(), EGRESS_HOST_FIELDS)
+        _emit_patch_event(serialized_host, existing_host.id, existing_host.canonical_facts.get("insights_id"))
+        return flask_json_response(serialized_host, 201)
```
Comment (rafaelcassau) [inline api/host.py:463]: I think return 200 is more accurate in this context, 201 is commonly used when a new resource was created.

Comment (Glutexo) [reply api/host.py:463]: That’s questionable. We may consider this being a new (virtual) check-in resource, not a direct modification of the Host. In that case, we should be using POST and not PUT. Even if we consider this to be a Host operation, PUT is still not appropriate – PATCH is.

File: api/host.py [hunk 9]
```diff
@@ -441,3 +445,29 @@ def _build_serialized_tags(host_list, search):
...
+    else:
+        return flask_json_response(
+            {
+                "detail": "No hosts match the provided facts.",
+                "status": 400,
+                "title": "Bad Request",
+                "type": "about:blank",
+            },
+            status=400,
```
Comment (rafaelcassau) [inline api/host.py:472]: maybe 404 fits better than 400 in this scenario.

Comment (Glutexo) [reply api/host.py:472]: Great point! Please see my comment above and tell me what you think! https://github.com/RedHatInsights/insights-host-inventory/pull/720#pullrequestreview-519590906

File: api/host.py (hunk 8 above)
Comment (rafaelcassau) [reply api/host.py:463]: yes, I'll argue about PATCH instead PUT because it just update a part of a host object instead of the whole object, I agree with you about semantic, I prefer to use PATCH but POST can be also considered.

**Comment by rafaelcassau** [comment]
> Is it correct to return _400 Bad Request_ when there is no matching host found? I’d expect _404 Not Found_ instead. I thought that _400 Bad Request_ should occur only if the request is malformed.
> 
> Gave this a second thought after reading https://stackoverflow.com/a/44915434/1307676. That use case is however slightly different from ours. Trying to wrap my head around our check-in, I imagine that the request consists of two parts: ⑴ matching the host and ⑵ writing the check-in data. No matter whether we consider this to be a new Check-in resource operation, or a direct Host update, it could be split into two requests: ⑴ would match the host and either redirect or bail out with _404 Not Found_. ⑵ would return _200 OK_ or _201 Created_ depending on the semantics.

I agree with you, it's a hard trade off, my point is, I agree 404 is related with the resource, but I always prefer to return more meaningful status code to the client, from the HTTP verb perspective 404 is not the best choice, but I think 400 is too generic, according with mozilla definition maybe 422 can be a better choice, what do you think?

https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/422

File: api/host.py (hunk 8 above)
Comment (Glutexo) [reply api/host.py:463]: It all depends on whether we consider the check-in to be a new resource.

- If yes: then it should be POST. We **create** a new object that internally manifests as a host stale timestamp update.
- If no: it should be PATCH, because PATCH provides a subset of object information, whereas PUT replaces the entire object.

After a discussion with Greg we both think that the former is correct. This is not an immediate patch, because we don’t provide new Host stale timestamp. We provide some information that the Inventory **uses** to calculate the new timestamp.

Does this make sense?

Comment (rafaelcassau) [reply api/host.py:463]: > It all depends on whether we consider the check-in to be a new resource.
> 
> * If yes: then it should be POST. We **create** a new object that internally manifests as a host stale timestamp update.
> * If no: it should be PATCH, because PATCH provides a subset of object information, whereas PUT replaces the entire object.
> 
> After a discussion with Greg we both think that the former is correct. This is not an immediate patch, because we don’t provide new Host stale timestamp. We provide some information that the Inventory **uses** to calculate the new timestamp.
> 
> Does this make sense?

yes, make sense, in this case we need to change the verb and we can maintain with 201 status code.

**Comment by Glutexo** [comment]
@rafaelcassau If we consider the Check-in to be bound to a specific Host, we can translate it (virtually or physically) into the match and update parts. Like that the Check-in create would be equivatlent to _POST /host/{host_id}/check-in_, in which case 404 Not Found would be the most appropriate choice.

I however agree that _422 Unprocessable Entity_ sounds nice and the description fits kind of. [Here](https://stackoverflow.com/a/44915434/1307676) though, one user says that this is a special-purpose status code introduced as a WebDAV extension and it is not thus a wise choice for a simple HTTP REST operation.

In conclusion, I’d go with

- _404 Not Found_ if Check-in is considered a new resource.
- _422 Unprocessable Entity_ or _400 Bad Request_ if Check-in is considered a patch on a Host.

As already mentioned above, I am more in favor of the former.

**Comment by kruai** [comment]
Y'all are leaving so many useful comments which would have been great before this merged 2 weeks ago :sweat_smile: 
I'm going to have to respond to most of this once I'm out of meetings; however, on the topic of POST vs PATCH, I'd lean towards PATCH. I have to squint really hard to see this thing as a "create" operation, especially since it's explicitly only an update on the code side

**Comment by kruai** [comment]
> One more comment to the schema. The new _CreateCheckIn_ schema is the only place in the
> whole specification where Canonical Facts is explicitly mentioned. We don’t expose this concept; for the outside world, these are first class properties of a Host object. Thus I don’t think it is correct to break this concept in the new endpoint.
> 
> I’d say, just put them in the root CreateCheckIn schema. _Flat is better than nested._ I personally like nested more, and even here, we could just call it like _host_ instead of _canonical_facts_, but it would be inconsistent.
> 
> ```yaml
> CreateCheckIn:
>   …
>   properties:
>     insights_id: …
>     suscription_manager_id: …
>     …
>     checkin_frequency: …
> ```
> 
> Of course, my suggestion to DRY the Canonical Facts property set to an extracted schema and use it here still applies. It would require to use the _allOf_ JSON Schema combinator.

I get what you mean - the host creation API is already deprecated, so I may as well delete the CreateHostIn schema, and just use the flat CreateCheckIn schema. Am I understanding you correctly?

**Comment by rafaelcassau** [comment]
> @rafaelcassau If we consider the Check-in to be bound to a specific Host, we can translate it (virtually or physically) into the match and update parts. Like that the Check-in create would be equivatlent to _POST /host/{host_id}/check-in_, in which case 404 Not Found would be the most appropriate choice.
> 
> I however agree that _422 Unprocessable Entity_ sounds nice and the description fits kind of. [Here](https://stackoverflow.com/a/44915434/1307676) though, one user says that this is a special-purpose status code introduced as a WebDAV extension and it is not thus a wise choice for a simple HTTP REST operation.
> 
> In conclusion, I’d go with
> 
> * _404 Not Found_ if Check-in is considered a new resource.
> * _422 Unprocessable Entity_ or _400 Bad Request_ if Check-in is considered a patch on a Host.
> 
> As already mentioned above, I am more in favor of the former.

agree
